packages/legend-graph/src/models/protocols/pure/v1/V1_PureGraphManager.ts

```diff
@@ -2016,6 +2036,237 @@ export class V1_PureGraphManager extends AbstractPureGraphManager {
     );
   }
 
+  async indexEntitiesWithDependencyIntoGraph<T extends V1_PackageableElement>(
+    graph: PureModel,
+    _entities: Entity[],
+    dependencyEntities: Map<string, Entity[]>,
+    filteredElementClazzes: GenericClazz<T>[],
+    // TODO we don't need this once protocol type is properly modularized
+    filteredProtocolTypes: string[],
+  ): Promise<V1_GraphBuilderInput[]> {
+    // remove unnecessary entities
+    Array.from(dependencyEntities.entries()).forEach(
+      ([dependencyKey, dEntities]) => {
+        const filteredEntities = dEntities.filter((e) =>
+          filteredProtocolTypes.includes(
+            (e.content as PlainObject<V1_PackageableElement>)._type as string,
+          ),
+        );
+        dependencyEntities.set(dependencyKey, filteredEntities);
+      },
+    );
+    const entities = _entities.filter((e) =>
+      filteredProtocolTypes.includes(
+        (e.content as PlainObject<V1_PackageableElement>)._type as string,
+      ),
+    );
+    const report = new GraphBuilderReport();
+    // build dependency pmcd models
+    const dependencyDataMap = new Map<string, V1_PureModelContextData>();
+    await Promise.all(
+      Array.from(dependencyEntities.entries()).map(([dependencyKey, value]) => {
+        const projectModelData = new V1_PureModelContextData();
+        dependencyDataMap.set(dependencyKey, projectModelData);
+        return V1_entitiesToPureModelContextData(
+          value,
+          projectModelData,
+
+          this.pluginManager.getPureProtocolProcessorPlugins(),
+        );
+      }),
+    );
+    const dependencyGraphBuilderInput: V1_GraphBuilderInput[] = Array.from(
+      dependencyDataMap.entries(),
+    ).map(([dependencyKey, dependencyData]) => ({
+      data: indexPureModelContextData(report, dependencyData, this.extensions),
+      model: graph.dependencyManager.getModel(dependencyKey),
+    }));
+    // build main pmcd
+    const data = new V1_PureModelContextData();
+    await V1_entitiesToPureModelContextData(
+      entities,
+      data,
+      this.pluginManager.getPureProtocolProcessorPlugins(),
+    );
+    const mainGraphBuilderInput: V1_GraphBuilderInput[] = [
+      {
+        model: graph,
+        data: indexPureModelContextData(report, data, this.extensions),
+      },
+    ];
+    const graphBuilderInput = [
+      ...dependencyGraphBuilderInput,
+      ...mainGraphBuilderInput,
+    ];
+    graphBuilderInput.forEach((e) => {
+      e.data.elements = e.data.elements.filter((el: V1_PackageableElement) =>
+        filteredElementClazzes.find((_clazz) => el instanceof _clazz),
+      );
+    });
+    await flowResult(this.initializeAndIndexElements(graph, graphBuilderInput));
+    return graphBuilderInput;
+  }
+
+  async buildCreateQueryLightGraph(
```

Comment:
Could we put a comment to keep my suggestion about optimizing this further, i.e. to tinker with the entities to omit parts we don't need? Thanks!